src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -1488,5 +1488,355 @@ public void nameof(string x)
             var option = TestOptions.ReleaseDll;
             CreateCompilation(source, options: option).VerifyDiagnostics();
         }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceInstanceMembersFromStaticMemberInNameof()
+        {
+            var source = @"
+System.Console.Write(C.M());
+public class C
+{
+    public C1 Property { get; }
+    public C1 Field { get; }
+    public event System.Action Event;
+    public static string M() => nameof(Property.Property) 
+        + "","" + nameof(Property.Field)
+        + "","" + nameof(Property.Method)
+        + "","" + nameof(Property.Event)
+        + "","" + nameof(Field.Property) 
+        + "","" + nameof(Field.Field)
+        + "","" + nameof(Field.Method)
+        + "","" + nameof(Field.Event)
+        + "","" + nameof(Event.Invoke)
+        ;
+}
+
+public class C1
+{
+    public int Property { get; }
+    public int Field;
+    public void Method(){}
+    public event System.Action Event;
+}";
+            CompileAndVerify(
+                source,
+                options: TestOptions.DebugExe,
+                parseOptions: TestOptions.RegularPreview,
+                expectedOutput: "Property,Field,Method,Event,Property,Field,Method,Event,Invoke").VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceInstanceMembersFromFieldInitializerInNameof()
+        {
+            var source = @"
+System.Console.Write(new C().S);
+public class C
+{
+    public string S { get; } = nameof(S.Length);
+}";
+            CompileAndVerify(
+                 source,
+                 options: TestOptions.DebugExe,
+                 parseOptions: TestOptions.RegularPreview,
+                 expectedOutput: "Length").VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceInstanceMembersFromAttributeInNameof()
+        {
+            var source = @"
+public class C
+{
+    [System.Obsolete(nameof(S.Length))]
```

Comment:
>nameof(S.Length) [](http://example.com/codeflow?start=21&length=16)

It should be possible to verify the value by referencing the property somewhere in code, I assume the value will be observable through diagnostics.